cmd/codegen/generator/go/templates/modules.go

```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
-				if err != nil {
-					return nil, fmt.Errorf("failed to parse struct tag: %w", err)
-				}
+		// arguments to WithArg
+		args := []Code{Lit(spec.graphqlName()), typeDef}
 
-				argTypeDef, err := ps.goTypeToAPIType(param.Type(), nil)
+		argOpts := []Code{}
+		if spec.description != "" {
+			argOpts = append(argOpts, Id("Description").Op(":").Lit(spec.description))
+		}
+		if spec.defaultValue != "" {
+			var jsonEnc string
+			if spec.baseType.String() == "string" {
+				enc, err := json.Marshal(spec.defaultValue)
 				if err != nil {
-					return nil, fmt.Errorf("failed to convert param type: %w", err)
-				}
-
-				argOptional := true
-				if tags != nil {
-					if tag, err := tags.Get("required"); err == nil {
-						argOptional = tag.Value() == "true"
-					}
+					return nil, fmt.Errorf("failed to marshal default value: %w", err)
 				}
+				jsonEnc = string(enc)
+			} else {
+				jsonEnc = spec.defaultValue
+			}
+			argOpts = append(argOpts, Id("DefaultValue").Op(":").Id("JSON").Call(Lit(jsonEnc)))
+		}
+		if len(argOpts) > 0 {
+			args = append(args, Id("FunctionWithArgOpts").Values(argOpts...))
+		}
 
-				// all values in a struct are optional
-				argTypeDef = argTypeDef.Dot("WithOptional").Call(Lit(argOptional))
+		fnDef = dotLine(fnDef, "WithArg").Call(args...)
+	}
 
-				// arguments to WithArg
-				argArgs := []Code{
-					Lit(param.Name()),
-					argTypeDef,
-				}
+	return fnDef, nil
+}
 
-				argOpts := []Code{}
+func (ps *parseState) parseParamSpecs(fn *types.Func) ([]paramSpec, error) {
+	sig := fn.Type().(*types.Signature)
+	params := sig.Params()
+	if params.Len() == 0 {
+		return nil, nil
+	}
 
-				if tags != nil {
-					// TODO: support this?
-					// if tag, err := tags.Get("name"); err == nil {
-					// 	def.Name = tag.Value()
-					// }
+	specs := make([]paramSpec, 0, params.Len())
 
-					if tag, err := tags.Get("doc"); err == nil {
-						argOpts = append(argOpts, Id("Description").Op(":").Lit(tag.Value()))
-					}
+	i := 0
+	if params.At(i).Type().String() == contextTypename {
+		spec, err := ps.parseParamSpecVar(params.At(i))
+		if err != nil {
+			return nil, err
+		}
+		specs = append(specs, spec)
 
-					if tag, err := tags.Get("default"); err == nil {
-						var jsonEnc string
-						if param.Type().String() == "string" {
-							enc, err := json.Marshal(tag.Value())
-							if err != nil {
-								return nil, fmt.Errorf("failed to marshal default value: %w", err)
-							}
-							jsonEnc = string(enc)
-						} else {
-							jsonEnc = tag.Value() // assume JSON encoded
-						}
-						argOpts = append(argOpts, Id("DefaultValue").Op(":").Id("JSON").Call(Lit(jsonEnc)))
-					}
-				}
+		i++
+	}
 
-				if len(argOpts) > 0 {
-					argArgs = append(argArgs, Id("FunctionWithArgOpts").Values(argOpts...))
-				}
+	fnDecl, err := ps.declForFunc(fn)
+	if err != nil {
+		return nil, err
+	}
 
-				fnDef = dotLine(fnDef, "WithArg").Call(argArgs...)
+	if i+1 == params.Len() {
```

Comment:
nit: took me a sec to grok, a quick comment about the case this is handling (inline struct) would be helpful